### ISSUE TITLE
Glitchy graphical flicker when resizing window

### ISSUE THREAD

**Issue by salty-horse** [garglk/garglk#679]
Using latest master on Linux.

When resizing the Gargoyle window, the contents of the window get smeared for a moment before redrawing correctly.

It started happening since commit 1ac6aeea54ffea08589263c95912515d7febf6d9 - Don't redraw directly inside of paint events.
![glitch](https://user-images.githubusercontent.com/28355/186929821-a300cc34-fe08-4928-8758-2b30df8c7ea6.png)

**Comment by cspiegel** [comment]
This seems to be because the canvas is being redrawn before it's resized, i.e. the window gets resized, draws the old canvas, the canvas gets resized, then is redrawn. If you resize vertically only, the artifacts don't occur, because the horizontal lines remain the same size.

There's a lot of moving parts to the redrawing, and I've not really done much with them, so I'll have to figure out how they go together. But this ought to be fairly easy to fix.

**Comment by angstsmurf** [comment]
Just another data point: On macOS, the sheared picture isn't just momentary. It appears when the user resizes the game window, and stays until it is resized again.

<img width="697" alt="Skärmavbild 2022-10-11 kl  02 08 44" src="https://user-images.githubusercontent.com/5066755/194970608-6fc86d9f-3549-41a1-a011-116c6e1c3aec.png">

**Comment by cspiegel** [comment]
Hm... ok, that's an issue.  The Mac display code is different than the Qt display code, but I modeled this part of the Qt code on the Mac code so I suppose it makes sense they'd exhibit similar behaviors.

It's really a pain dealing with the resizing signals and threading all that information through at the right time, but I'll try to get a handle on it.

**Comment by Nephar** [comment]
I can reproduce this issue without window resizing(the terp launches looking like below):

<img width="751" alt="Screen Shot 2023-01-14 at 10 26 38" src="https://user-images.githubusercontent.com/122389007/212477087-2dd788c1-d995-44e3-80f4-ce461ceaad6f.png">

To reproduce it, only 2 changes to the default garglk.ini are needed.
1. increase cols to 61
2. increase monosize to 18

Both changes are for global settings. I believe with that column width, monosize between 17.6-19.2 causes this screen. This bug might only present itself under Mojave, since that is what I am using.

**Comment by angstsmurf** [comment]
I can confirm that the same thing happens on macOS 13.1 Ventura if I make those changes to the default settings.

**Comment by cspiegel** [comment]
OK, I can reproduce it on Mac, but not anywhere else; interesting, since the _resize_ glitch exists on Qt platforms, which is also related to column size (if you resize vertically, it doesn't happen).

The fact that a config setting can cause this is more urgent than just the resizing, so I'll have to try to track it down now.  Thanks for the detailed info on reproducing!

**Comment by angstsmurf** [comment]
I made a pull request which seems to fix this on macOS.

**Comment by cspiegel** [comment]
It does look like this is a different bug on both Qt and Mac: I have a tentative workaround for Qt which does nothing for the Mac. If the source of the Mac bug can't be pinned down I'll apply the proposed "no odds" fix, which is a really clever way to paper over the problem!

To update my earlier statement, on Qt, the canvas is being resized and displayed _before_ the canvas is written to, i.e. the canvas contains old data (with the wrong width), and that's why it's displayed wrong. The obvious fix is to redraw the canvas right after resizing it, which appears to work, generally, though may have unintended consequences.  I'm testing that.

On Mac, it's a different story. The canvas data always looks right. You can see this by dumping it to a PPM file in `addFrame` in `launchmac.mm`.  At the end of the function, just add:

```
std::ofstream f("out.ppm");
f << "P6\n" << width << " " << height << "\n255\n";
f.write((const char *)[frame bytes], width * height * 3);
```

The data's good, but when it gets drawn... I'm unfamiliar with the Mac display model, so I'm just doing lots of trial and error here.  At any rate, if I can't pin this down soon I'll add your fix, and kick this down the road a bit.

Thank you for working on this!

**Comment by Nephar** [comment]
I like the "no odds" fix too. The nice thing is about this fix is I don't even have to wait for the release, just updated my config to 62/18 for cols/monosize, and it's good to go.

**Comment by angstsmurf** [comment]
[This](https://stackoverflow.com/questions/54165935/opengl-displayed-image-sheared) made me try to simply replace `glPixelStorei(GL_UNPACK_CLIENT_STORAGE_APPLE, GL_TRUE);` at line 81 in `launchmac.mm` with `glPixelStorei(GL_UNPACK_ALIGNMENT, GL_TRUE);` (i.e. replace GL_UNPACK_CLIENT_STORAGE_APPLE  with GL_UNPACK_ALIGNMENT) which seems to fix it.

**Comment by cspiegel** [comment]
> [This](https://stackoverflow.com/questions/54165935/opengl-displayed-image-sheared) made me try to simply replace `glPixelStorei(GL_UNPACK_CLIENT_STORAGE_APPLE, GL_TRUE);` at line 81 in `launchmac.mm` with `glPixelStorei(GL_UNPACK_ALIGNMENT, GL_TRUE);` (i.e. replace GL_UNPACK_CLIENT_STORAGE_APPLE with GL_UNPACK_ALIGNMENT) which seems to fix it.
> 
> EDIT: Reading that answer again, it seems the second parameter should be 1, not GL_TRUE, as it can be set to one of four different byte values.

I think you nailed it!  I can't reproduce the problem at all now.  Thank you so much for tracking that down.  If you want to create a pull request with that, I'll merge it; otherwise, I can create a pull request with this plus my fix for the Qt rendering side of things.

**Comment by angstsmurf** [comment]
Go ahead. I'll close the other one.

**Comment by cspiegel** [comment]
Ok, fixed in #732. I'll leave this open a bit so anybody who's interested can test and ensure things are up to snuff, but it's all looking good for me.

Thank you again for finding this fix!

**Comment by cspiegel** [comment]
No problems in over a year; closing.
